crates/proxy-server/src/main.rs

```diff
@@ -1,81 +1,290 @@
+            );
+        }
+
+        if self.is_legacy_mode() {
+            if self.listen.is_empty() {
+                anyhow::bail!(
+                    "No listeners defined! Use --listen to specify at least one, \
+                     or use --policy for Lua-based configuration."
+                );
+            }
+
+            if self.tls_private_key.is_some() != self.tls_certificate.is_some() {
+                anyhow::bail!(
+                    "--tls-certificate and --tls-private-key must be specified together, or not at all"
+                );
+            }
+        }
 
-    if opts.listen.is_empty() {
-        anyhow::bail!("No listeners defined! use the --listen option to specify at least one!");
+        Ok(())
     }
+}
+
+fn main() -> anyhow::Result<()> {
+    let opts = Opt::parse();
+    opts.validate()?;
 
     let (_no_file_soft, no_file_hard) = getrlimit(Resource::RLIMIT_NOFILE)?;
     setrlimit(Resource::RLIMIT_NOFILE, no_file_hard, no_file_hard).context("setrlimit NOFILE")?;
 
+    kumo_server_common::panic::register_panic_hook();
+
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .on_thread_park(kumo_server_memory::purge_thread_cache)
+        .build()
+        .unwrap()
+        .block_on(async move { run(opts).await })
+}
+
+async fn perform_init() -> anyhow::Result<()> {
+    tracing::info!("Version is {}", version_info::kumo_version());
+    let mut config = config::load_config().await?;
+
+    let proxy_init_sig = CallbackSignature::<(), ()>::new("proxy_init");
+
+    config
+        .async_call_callback(&proxy_init_sig, ())
+        .await
+        .context("in proxy_init event")?;
+    config.put();
+
+    Ok(())
+}
+
+async fn signal_shutdown() {
+    tracing::info!("shutting down");
+}
+
+async fn run(opts: Opt) -> anyhow::Result<()> {
+    kumo_server_runtime::assign_main_runtime(tokio::runtime::Handle::current());
+
+    if opts.is_legacy_mode() {
+        // Legacy mode: use CLI options directly without Lua
+        run_legacy(opts).await
+    } else {
+        // Modern mode: use Lua policy file
+        let policy = opts
+            .policy
+            .unwrap_or_else(|| PathBuf::from("/opt/kumomta/etc/proxy/init.lua"));
+
+        StartConfig {
+            logging: LoggingConfig {
+                log_dir: opts.diag_log_dir.clone(),
+                diag_format: opts.diag_format,
+                filter_env_var: "KUMO_PROXY_LOG",
+                default_filter:
+                    "proxy_server=info,kumo_server_common=info,kumo_server_runtime=info",
+            },
+            lua_funcs: &[kumo_server_common::register, mod_proxy::register],
+            policy: &policy,
+        }
+        .run(perform_init(), signal_shutdown())
+        .await
+    }
+}
+
+/// Run in legacy mode using CLI options (backwards compatibility)
+async fn run_legacy(opts: Opt) -> anyhow::Result<()> {
```

Comment:
I would love to avoid this code duplication!

My suggestion is that when the user doesn't provide a path to a config script, that we create one based on the command line arguments:

* Use `tempfile::NamedTempFile::with_prefix("kumo-proxy-config")` to create a temporary file
* We write out to it based on what we see in `opts`:

```lua
local kumo = require 'kumo'

kumo.on('proxy_init', function()
```

Then loop over the `opts.listen` and emit appropriate listener entries:

```
  kumo.start_proxy_listener {
    listen = LISTENER,
    -- no_splice and timeout here too   
  }
```

and then close it out:

```
end)
```

With that populated, we can then pass in that path via `StartConfig::policy`.